OpenRA.Mods.Common/Activities/FindResources.cs

```diff
@@ -64,44 +103,46 @@ public override Activity Tick(Actor self)
-
-				harv.LastSearchFailed = true;
-
-				var unblockCell = harv.LastHarvestedCell ?? (self.Location + harvInfo.UnblockCell);
-				var moveTo = mobile.NearestMoveableCell(unblockCell, 2, 5);
-				self.QueueActivity(mobile.MoveTo(moveTo, 1));
-
-				foreach (var n in self.TraitsImplementing<INotifyHarvesterAction>())
-					n.MovingToResources(self, moveTo, this);
-
-				self.SetTargetLine(Target.FromCell(self.World, moveTo), Color.Gray, false);
-				var randFrames = self.World.SharedRandom.Next(100, 175);
-
-				// Avoid creating an activity cycle
-				var next = NextInQueue;
-				NextInQueue = null;
-				return ActivityUtils.SequenceActivities(self, next, new Wait(randFrames), this);
+				// If no resources are found near the current field, search near the refinery instead.
+				// If that doesn't help, give up for now.
+				if (harv.LastHarvestedCell != null)
+					harv.LastHarvestedCell = null;
+				else
+					harv.LastSearchFailed = true;
+
+				var lastproc = harv.LastLinkedProc ?? harv.LinkedProc;
+				if (lastproc != null && !lastproc.Disposed)
+				{
+					var deliveryLoc = lastproc.Location + lastproc.Trait<IAcceptResources>().DeliveryOffset;
+					if (self.Location == deliveryLoc && harv.IsEmpty)
+					{
+						// Get out of the way:
+						var unblockCell = deliveryLoc + harv.Info.UnblockCell;
+						var moveTo = mobile.NearestMoveableCell(unblockCell, 1, 5);
+						self.SetTargetLine(Target.FromCell(self.World, moveTo), Color.Green, false);
+						QueueChild(self, mobile.MoveTo(moveTo, 1), true);
+					}
+				}
+
+				return this;
 			}
-			else
-			{
-				// Attempt to claim the target cell
-				if (!claimLayer.TryClaimCell(self, closestHarvestablePosition.Value))
-					return ActivityUtils.SequenceActivities(self, new Wait(25), this);
 
-				harv.LastSearchFailed = false;
+			// Attempt to claim the target cell
+			if (!claimLayer.TryClaimCell(self, closestHarvestablePosition.Value))
+			{
+				QueueChild(self, new Wait(25), true);
+				return this;
+			}
 
-				// If not given a direct order, assume ordered to the first resource location we find:
-				if (!harv.LastOrderLocation.HasValue)
-					harv.LastOrderLocation = closestHarvestablePosition;
+			harv.LastSearchFailed = false;
 
-				foreach (var n in self.TraitsImplementing<INotifyHarvesterAction>())
-					n.MovingToResources(self, closestHarvestablePosition.Value, this);
+			foreach (var n in self.TraitsImplementing<INotifyHarvesterAction>())
+				n.MovingToResources(self, closestHarvestablePosition.Value, new FindResources(self));
 
-				self.SetTargetLine(Target.FromCell(self.World, closestHarvestablePosition.Value), Color.Red, false);
-				return ActivityUtils.SequenceActivities(self, mobile.MoveTo(closestHarvestablePosition.Value, 1), new HarvestResource(self), this);
-			}
+			self.SetTargetLine(Target.FromCell(self.World, closestHarvestablePosition.Value), Color.Red, false);
+			QueueChild(self, mobile.MoveTo(closestHarvestablePosition.Value, 1), true);
```

Comment:
A good followup improvement for a future PR will be to move this and the `MovingToResources` notification above inside `HarvestResource`

Reply:
Sounds reasonable. This refactor can be included when fixing #14752.